src/main/clojure/aerospike_clj/collections.clj

```diff
@@ -0,0 +1,19 @@
+(ns aerospike-clj.collections
+  (:import (java.util ArrayList Collection Collections List)
+           (java.util.function Consumer)))
+
+(defn ->list
+  "Returns a new [[java.util.List]] containing the result of applying `mapper-fn` to each item in `col`.
+   Returns an unmodifiable list.
+   *Note*: This will usually be faster than `(mapv mapper-fn col)` because:
+           - This function allocates a new [[java.util.ArrayList]] in the exact `(.size col)` size, and then
+             fills it with the mapped values.
+           - If the underlying collection is not a Clojure sequence, then `mapv` will first convert it
+             to a Clojure sequence and then map over it. This function will not do that."
+  ^List [mapper-fn ^Collection col]
+  (let [res (ArrayList. (.size col))]
+    (.forEach col
+              (reify Consumer
+                (accept [_ item]
+                  (.add res (mapper-fn item)))))
```

Comment:
Is the `reify` call optimized by the Clojure compile to happen once for the namespace (and then `res` is bound in the lexical scope of `->list`), once for each invocation of `->list` or once for every item that's iterated in `.forEach`?

Reply:
We can look into it using [clj-java-decompiler](https://github.com/clojure-goes-fast/clj-java-decompiler), and see that a reified instance is created once per invocation of `->list`, and that `res` and `mapper-fn` are passed as ctor params to it.

```clojure
(decompile
  (defn ->list
    "Returns a new [[java.util.List]] containing the result of applying `mapper-fn` to each item in `col`.
     Returns an unmodifiable list.
     *Note*: This will usually be faster than `(mapv mapper-fn col)` because:
             - This function allocates a new [[java.util.ArrayList]] in the exact `(.size col)` size, and then
               fills it with the mapped values.
             - If the underlying collection is not a Clojure sequence, then `mapv` will first convert it
               to a Clojure sequence and then map over it. This function will not do that."
    ^List [mapper-fn ^Collection col]
    (let [res (ArrayList. (.size col))]
      (.forEach col
                (reify Consumer
                  (accept [_ item]
                    (.add res (mapper-fn item)))))
      (Collections/unmodifiableList res))))

// Decompiling class: aerospike_clj/collections$__GT_list
package aerospike_clj;

import java.util.function.*;
import java.util.*;
import clojure.lang.*;

public final class collections$__GT_list extends AFunction
{
    public static final AFn const__4;
    
    public static Object invokeStatic(final Object mapper_fn, final Object col) {
        final Object res = new ArrayList(((Collection)col).size());
        ((Iterable)col).forEach((Consumer)new collections$$reify__2621(null, mapper_fn, res).withMeta((IPersistentMap)const__4));
        return Collections.unmodifiableList((List<?>)res);
    }
    
    @Override
    public Object invoke(final Object mapper_fn, final Object col) {
        return invokeStatic(mapper_fn, col);
    }
    
    static {
        const__4 = (AFn)RT.map(RT.keyword(null, "line"), 13L, RT.keyword(null, "column"), 17L);
    }
}


// Decompiling class: aerospike_clj/collections$$reify__2621
package aerospike_clj;

import java.util.function.*;
import java.util.*;
import clojure.lang.*;

public final class collections$$reify__2621 implements Consumer, IObj
{
    final IPersistentMap __meta;
    Object mapper_fn;
    Object res;
    
    public collections$$reify__2621(final IPersistentMap _meta, final Object mapper_fn, final Object res) {
        this.__meta = _meta;
        this.mapper_fn = mapper_fn;
        this.res = res;
    }
    
    public collections$$reify__2621(final Object o, final Object o2) {
        this(null, o, o2);
    }
    
    @Override
    public IPersistentMap meta() {
        return this.__meta;
    }
    
    @Override
    public IObj withMeta(final IPersistentMap persistentMap) {
        return new collections$$reify__2621(persistentMap, this.mapper_fn, this.res);
    }
    
    @Override
    public void accept(final Object item) {
        final Boolean b = ((ArrayList)this.res).add(((IFn)this.mapper_fn).invoke(item)) ? Boolean.TRUE : Boolean.FALSE;
    }
}

=> nil
```